In [1]:
!pip install geopandas

In [2]:
import calitp
from calitp.tables import tbl
from calitp import get_engine
from siuba import *

import pandas as pd
import numpy as np
import geopandas as gpd

In [3]:
engine = get_engine()
connection = engine.connect()

In [4]:
from calitp import query_sql

In [5]:
query_sql("select * from views.mst_ridership")

participant_id              littlepay_transaction_id device_id  \
0                  sacrt  50e5c916-3f88-48f9-ae2d-3adbdc70a1c0  17F313D3   
1                  sacrt  b20c4c5a-5bf7-412d-8aa1-f767e9fdaf57  17F313D3   
2                  sacrt  f18e39e2-3eb5-4d48-8009-e25565b9fae9  17F313D3   
3                  sacrt  ee2c45ee-9d9b-43f9-ba33-aa8dba13f967  17F313D3   
4                  sacrt  047b4251-321f-44ee-9e3f-b6d602783bbd  17F313D3   
...                  ...                                   ...       ...   
39272  clean-air-express  dc20e795-261b-41c6-98a2-a620c85fa229     61005   
39273  clean-air-express  04d94e51-0384-46ff-9c3f-0722cb8191cd     61005   
39274  clean-air-express  26da56c2-185d-445a-b9fe-f7e24b449cca     61005   
39275  clean-air-express  60846f92-8e6c-49e8-9f29-23264d56384a     61005   
39276  clean-air-express  157d4863-6ce6-4147-88d0-ba1f3e434ff3     61005   

                            micropayment_id    type transaction_outcome  \
0      f8136ba0-5f66-410c-aa2e-736ae14fbbf6  single               allow   
1      698c2b29-14a3-4a25-b20a-f836b4cd9412  single               allow   
2      269d7920-9ae3-491c-be7c-f4061d0b9658  single               allow   
3      fe8a234f-26d4-4929-9c4e-9c2c5e89e9c6  single               allow   
4      54ef76b3-ead4-4cdc-90e4-5b5fe98cb624  single               allow   
...                                     ...     ...                 ...   
39272  fae9164c-9de8-414a-95d4-28234e12f7ae  single               allow   
39273  5eaa4fad-53b5-4e1d-a012-a5d83832ecf2  single               allow   
39274  701f04b8-737d-4daf-b6c1-cd3255bc8165  single               allow   
39275  be1b0cc0-f74c-4acb-b3c7-7fe566219b63  single               allow   
39276  ebe1340a-b104-4cb1-abfb-ec44343da5cf  single               allow   

      transaction_date_time_utc transaction_date_time_pacific location_id  \
0      2021-09-28T17:26:29.000Z           2021-09-28 10:26:29        7018   
1      2021-10-25T15:29:17.000Z           2021-10-25 08:29:17        7026   
2      2021-10-25T15:35:42.000Z           2021-10-25 08:35:42        7029   
3      2021-09-28T16:03:41.000Z           2021-09-28 09:03:41        7043   
4      2021-09-30T22:58:45.000Z           2021-09-30 15:58:45        7051   
...                         ...                           ...         ...   
39272  2021-08-05T00:20:09.000Z           2021-08-04 17:20:09     1679031   
39273  2021-08-10T12:55:57.000Z           2021-08-10 05:55:57     1689413   
39274  2021-08-09T12:57:28.000Z           2021-08-09 05:57:28     1689413   
39275  2021-08-09T12:53:45.000Z           2021-08-09 05:53:45     1689413   
39276  2021-08-11T12:55:24.000Z           2021-08-11 05:55:24     1689413   

                   location_name route_id            latitude  \
0          Broadway Station (EB)  Route Z   38.55976104736328   
1            Florin Station (WB)  Route Z   38.49806594848633   
2      City College Station (WB)  Route Z   38.54241943359375   
3        Royal Oaks Station (EB)  Route Z   38.60477828979492   
4       29th Street Station (EB)  Route Z    38.5640754699707   
...                          ...      ...                 ...   
39272              CtgePebCast O     None  34.428949499999995   
39273                     Lompoc     None  34.637450333333334   
39274                     Lompoc     None  34.637398499999996   
39275                     Lompoc     None   34.63740433333333   
39276                     Lompoc     None   34.63738316666667   

                 longitude          vehicle_id route_long_name  \
0       -121.4881820678711                 101            None   
1      -121.47150421142578                 101            None   
2      -121.48528289794922                 101            None   
3      -121.44329833984375                 101            None   
4      -121.46965789794922                 101            None   
...                    ...                 ...             ...   
39272  -119.72366933333332  

In [6]:
df = pd.read_sql_query("select * from views.mst_ridership WHERE participant_id='mst'", connection)

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32873 entries, 0 to 32872
Data columns (total 19 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   participant_id                 32873 non-null  object        
 1   littlepay_transaction_id       32873 non-null  object        
 2   device_id                      32873 non-null  object        
 3   micropayment_id                32873 non-null  object        
 4   type                           32873 non-null  object        
 5   transaction_outcome            32873 non-null  object        
 6   transaction_date_time_utc      32873 non-null  object        
 7   transaction_date_time_pacific  32873 non-null  datetime64[ns]
 8   location_id                    32873 non-null  object        
 9   location_name                  32873 non-null  object        
 10  route_id                       32873 non-null  object        
 11  latitude       

In [8]:
df.head()

participant_id              littlepay_transaction_id device_id  \
0            mst  09535eb0-da27-4e74-9336-d090cbf18a17  17F12923   
1            mst  98a770a5-3f45-45cb-a640-fac02a218021  17F12923   
2            mst  4a8f6396-16d2-4f8b-9a16-9b2f9acc9354  17F12923   
3            mst  c5e9c00e-61f5-45c2-ab65-ed699b272643  17F12923   
4            mst  41f53f52-cf04-45f1-938a-910f41eebbff  17F12923   

                        micropayment_id    type transaction_outcome  \
0  36382bb2-60ce-449e-879d-dabbb9046570  single               allow   
1  d5e86cca-95d1-441d-80ea-e451d688e960  single               allow   
2  e5dfdb06-785e-48ba-936c-33ae62fae0e2  single               allow   
3  5ae7dc1c-e4cb-4b5d-afe8-1b55129d8336  single               allow   
4  71e85a22-9236-484a-9c75-9e8e2784ff0c  single               allow   

  transaction_date_time_utc transaction_date_time_pacific   location_id  \
0  2021-07-14T16:59:43.000Z           2021-07-14 09:59:43  039026150001   
1  2021-05-13T23:53:51.000Z           2021-05-13 16:53:51  039026150001   
2  2021-07-14T16:22:47.000Z           2021-07-14 09:22:47  039026150001   
3  2021-11-17T22:35:20.000Z           2021-11-17 14:35:20  039026150001   
4  2021-10-29T20:08:05.000Z           2021-10-29 13:08:05  039026150001   

  location_name route_id           latitude            longitude vehicle_id  \
0        Stop A  Route Z  36.32857894897461  -121.23577880859375        946   
1        Stop A  Route Z   36.6993522644043  -121.80551147460938        946   
2        Stop A  Route Z      36.3212890625  -121.24269104003906        946   
3        Stop A  Route Z  36.31319808959961  -121.24486541748047        946   
4        Stop A  Route Z  36.32319641113281  -121.25514221191406        946   

  route_long_name route_short_name             charge_type  charge_amount  \
0            None             None  complete_variable_fare            1.5   
1            None             None               flat_fare            3.5   
2            None             None  complete_variable_fare            1.5   
3            None             None  complete_variable_fare            1.5   
4            None             None  complete_variable_fare            1.5   

   nominal_amount  
0             1.5  
1             3.5  
2             1.5  
3             1.5  
4             1.5

Route Type	Regular Fare	Discount Fare
Local	    $1.50	        $.75
Primary	    $2.50	        $1.25
Regional	$3.50	        $1.75
Commuter	$12.00	        $6.00

Distance	                Regular Contactless Fare
Up to 2.7 miles	            $1.50
Between 2.7 and 14 miles	$2.50
14 – 50 miles	            $3.50
50 – 77 miles	            $7.00
77 miles or more	        $10.00

In [9]:
df['charge_type'].value_counts()

complete_variable_fare    28056
flat_fare                  4817
Name: charge_type, dtype: int64

In [10]:
df['charge_amount'].max()

10.0

No one was charged the 12 regular commuter fare; capped out at 10.

In [11]:
df['charge_amount'].min()

0.0

In [12]:
df['route_id'].value_counts()

020        5906
902        3775
018        2982
002        2792
901        2479
041        2407
024        2081
023        2025
Route Z    1188
001        1144
044         854
045         746
049         725
075         697
029         510
011         438
028         300
094         297
022         285
093         271
016         244
003         135
095         111
021          95
061          87
025          70
091          67
026          46
019          32
084          29
048          18
092          15
007          14
047           4
042           4
Name: route_id, dtype: int64

** are the 252 0.00 fares actually no fares or missing data?

In [13]:
df['charge_amount'].value_counts()

1.5     14796
2.5     11086
3.5      6569
0.0       252
10.0       82
8.5        18
1.0        16
7.5        12
2.0        10
7.0        10
0.5         8
6.5         6
5.0         4
6.0         2
3.0         2
Name: charge_amount, dtype: int64

*df['charge_amount'].equals(df['nominal_amount'])

*comparison_column = np.where(df["charge_amount"] == df["nominal_amount"], True, False)
*print(comparison_column)

*df["equal"] = comparison_column
*print(df)

In [14]:
df['higher_fare'] = df['charge_amount'].gt(df['nominal_amount'])
df['lower_fare'] = df['charge_amount'].lt(df['nominal_amount'])
df['equal_fare'] = df['charge_amount'].eq(df['nominal_amount'])

In [15]:
df.head()

participant_id              littlepay_transaction_id device_id  \
0            mst  09535eb0-da27-4e74-9336-d090cbf18a17  17F12923   
1            mst  98a770a5-3f45-45cb-a640-fac02a218021  17F12923   
2            mst  4a8f6396-16d2-4f8b-9a16-9b2f9acc9354  17F12923   
3            mst  c5e9c00e-61f5-45c2-ab65-ed699b272643  17F12923   
4            mst  41f53f52-cf04-45f1-938a-910f41eebbff  17F12923   

                        micropayment_id    type transaction_outcome  \
0  36382bb2-60ce-449e-879d-dabbb9046570  single               allow   
1  d5e86cca-95d1-441d-80ea-e451d688e960  single               allow   
2  e5dfdb06-785e-48ba-936c-33ae62fae0e2  single               allow   
3  5ae7dc1c-e4cb-4b5d-afe8-1b55129d8336  single               allow   
4  71e85a22-9236-484a-9c75-9e8e2784ff0c  single               allow   

  transaction_date_time_utc transaction_date_time_pacific   location_id  \
0  2021-07-14T16:59:43.000Z           2021-07-14 09:59:43  039026150001   
1  2021-05-13T23:53:51.000Z           2021-05-13 16:53:51  039026150001   
2  2021-07-14T16:22:47.000Z           2021-07-14 09:22:47  039026150001   
3  2021-11-17T22:35:20.000Z           2021-11-17 14:35:20  039026150001   
4  2021-10-29T20:08:05.000Z           2021-10-29 13:08:05  039026150001   

  location_name  ...            longitude vehicle_id route_long_name  \
0        Stop A  ...  -121.23577880859375        946            None   
1        Stop A  ...  -121.80551147460938        946            None   
2        Stop A  ...  -121.24269104003906        946            None   
3        Stop A  ...  -121.24486541748047        946            None   
4        Stop A  ...  -121.25514221191406        946            None   

  route_short_name             charge_type charge_amount nominal_amount  \
0             None  complete_variable_fare           1.5            1.5   
1             None               flat_fare           3.5            3.5   
2             None  complete_variable_fare           1.5            1.5   
3             None  complete_variable_fare           1.5            1.5   
4             None  complete_variable_fare           1.5            1.5   

   higher_fare  lower_fare  equal_fare  
0        False       False        True  
1        False       False        True  
2        False       False        True  
3        False       False        True  
4        False       False        True  

[5 rows x 22 columns]

In [16]:
df['equal_fare'].value_counts()

True     32671
False      202
Name: equal_fare, dtype: int64

In [17]:
df['higher_fare'].value_counts()

False    32873
Name: higher_fare, dtype: int64

In [18]:
df['lower_fare'].value_counts()

False    32671
True       202
Name: lower_fare, dtype: int64

So this means that fare capping worked, yes? No one was charged a higher amount than the nominal fare? 

In [19]:
df['charge_type'].value_counts()

complete_variable_fare    28056
flat_fare                  4817
Name: charge_type, dtype: int64

In [20]:
(
tbl.views.mst_ridership()
    >> filter(_.charge_amount < _.nominal_amount, _.charge_type == "flat_fare")
    >> summarize(n_fares = _.charge_type.count()
                )
)

# Source: lazy query
# DB Conn: Engine(bigquery://cal-itp-data-infra/?maximum_bytes_billed=5000000000)
# Preview:
   n_fares
0      996
# .. may have more rows

In [21]:
(
tbl.views.mst_ridership()
    >> filter(_.charge_amount < _.nominal_amount, _.charge_type == "complete_variable_fare")
    >> summarize(n_fares = _.charge_type.count()
                )
)

# Source: lazy query
# DB Conn: Engine(bigquery://cal-itp-data-infra/?maximum_bytes_billed=5000000000)
# Preview:
   n_fares
0      202
# .. may have more rows

Here are the things I need to make: 
* TABLE: number of people paying flat fare/contactless far
* TABLE: Descriptive stats on differences in fares paid by people using flat v. contactless
** (transaction type by time of day, day of week, route)

*November 29: Ok, so! 
27,926 paid a variable fare
4,781 paid a flat fare
    202 of the 27,926 people paid a lower fare